src/test/shared/vscode/fakeEnv.ts

```diff
@@ -0,0 +1,46 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Clipboard, Env } from '../../../shared/vscode/env'
+
+export interface FakeEnvOptions {
+    clipboard?: FakeClipboard | FakeClipboardOptions
+}
+
+export class FakeEnv implements Env {
+    private readonly _clipboard: FakeClipboard
+
+    public get clipboard(): FakeClipboard {
+        return this._clipboard
+    }
+
+    public constructor({ clipboard = new FakeClipboard() }: FakeEnvOptions = {}) {
+        this._clipboard = clipboard instanceof FakeClipboard ? clipboard : new FakeClipboard(clipboard)
```

Comment:
if clipboard isn't what you expected it to be, consider throwing instead of masking the issue with a different object, because the test was created incorrectly.

This comment applies to fakeWindow, and anywhere else that does this.

Reply:
This is an attempt to implement a pattern where the user can pass either an instance of the class OR just the constructor args.

This reduces the verbosity with nested faking.

```ts
new FakeWindow({ 
  inputBox: { input: bucketName },
  dialog: { saveSelection: saveLocation },
})
```

versus
```ts
new FakeWindow({ 
  inputBox: new FakeInputBox({ input: bucketName }),
  dialog: new FakeDialog({ saveSelection: saveLocation }),
})
```

Reply:
Ok I follow that, thanks for explaining.